File: android/src/main/kotlin/com/apparence/camerawesome/CameraPreviewTest.java
package com.apparence.camerawesome;

import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CaptureRequest;
import android.util.Size;
import android.view.Surface;

import com.apparence.camerawesome.models.FlashMode;
import com.apparence.camerawesome.surface.SurfaceFactory;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import io.flutter.view.TextureRegistry;

import static android.hardware.camera2.CaptureRequest.*;
import static android.hardware.camera2.CaptureRequest.FLASH_MODE;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class CameraPreviewTest {

    @Mock
    CameraDevice cameraDeviceMock;

    @Mock
    Builder captureRequestBuilder;

    @Mock
    SurfaceFactory surfaceFactoryMock;

    @Mock
    Surface surfaceMock;

    CameraSession cameraSession;

    CameraPreview cameraPreview;

    @Before
    public void setUp() throws Exception {
        reset(captureRequestBuilder);
        when(surfaceFactoryMock.build(any(Size.class))).thenReturn(surfaceMock);
        when(cameraDeviceMock.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)).thenReturn(captureRequestBuilder);
        cameraSession = new CameraSession();
        cameraPreview = new CameraPreview(
                cameraSession,
                null,
                surfaceFactoryMock,
                null,
                false);
        cameraSession.setOnCaptureSessionListenerList(
                Collections.<CameraSession.OnCaptureSession>singletonList(cameraPreview));
    }

    @Test
    public void createPreviewSession() throws CameraAccessException {
        cameraPreview.setPreviewSize(640, 480);
        cameraPreview.createCameraPreviewSession(cameraDeviceMock);
        Assert.assertNotNull(cameraPreview.getPreviewRequest());
        // Flash is disabled by default
        verify(captureRequestBuilder, times(1))
                .set(CaptureRequest.FLASH_MODE, CaptureRequest.FLASH_MODE_OFF);
        // AutoFocus is activated by default
        verify(captureRequestBuilder, times(1))
                .set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);
        // Android preview stays on portrait mode
        verify(captureRequestBuilder, times(1))
                .set(CaptureRequest.JPEG_ORIENTATION, 270);
        // check surfaces have been set
        Assert.assertEquals(cameraSession.getSurfaces().size(), 1);
        Assert.assertEquals(cameraSession.getCameraDevice(), cameraDeviceMock);
        Assert.assertNotNull(cameraSession.getCaptureSession());
    }

    @Test
    public void setFocusWithNoPreview()  {
        cameraPreview.setAutoFocus(true);
        verify(captureRequestBuilder, never()).set(eq(CaptureRequest.CONTROL_AF_MODE), Mockito.anyInt());
    }

    @Test
    public void setFocusWithPreview() throws CameraAccessException {
        cameraPreview.setPreviewSize(640, 480);
        cameraPreview.createCameraPreviewSession(cameraDeviceMock);
        reset(captureRequestBuilder);
        cameraPreview.setAutoFocus(true);
        verify(captureRequestBuilder, atLeastOnce()).set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);
        reset(captureRequestBuilder);
        cameraPreview.setAutoFocus(false);
        verify(captureRequestBuilder, atLeastOnce()).set(eq(CaptureRequest.CONTROL_AF_MODE), eq(CONTROL_AF_MODE_OFF));
    }

    @Test
    public void setFlashMode() throws CameraAccessException {
        cameraPreview.setPreviewSize(640, 480);
        cameraPreview.createCameraPreviewSession(cameraDeviceMock);
        reset(captureRequestBuilder);
        cameraPreview.setFlashMode(FlashMode.NONE);
        verify(captureRequestBuilder, atLeastOnce()).set(FLASH_MODE, CaptureRequest.FLASH_MODE_OFF);
        reset(captureRequestBuilder);
        cameraPreview.setFlashMode(FlashMode.AUTO);
        verify(captureRequestBuilder, atLeastOnce()).set(eq(CaptureRequest.FLASH_MODE), eq(FLASH_MODE_SINGLE));
        cameraPreview.setFlashMode(FlashMode.ALWAYS);
        verify(captureRequestBuilder, atLeastOnce()).set(eq(CaptureRequest.FLASH_MODE), eq(FLASH_MODE_TORCH));
    }
}

File: android/src/main/kotlin/com/apparence/camerawesome/CameraPreviewTest.kt
//package com.apparence.camerawesome
//
//import android.hardware.camera2.CameraAccessException
//import android.hardware.camera2.CameraDevice
//import android.hardware.camera2.CameraMetadata
//import android.hardware.camera2.CaptureRequest
//import android.util.Size
//import android.view.Surface
//import com.apparence.camerawesome.models.FlashMode
//import com.apparence.camerawesome.surface.SurfaceFactory
//import org.junit.Assert
//import org.junit.Before
//import org.junit.Test
//import org.junit.runner.RunWith
//import java.lang.Exception
//
//@RunWith(MockitoJUnitRunner::class)
//class CameraPreviewTest {
//    @Mock
//    var cameraDeviceMock: CameraDevice? = null
//
//    @Mock
//    var captureRequestBuilder: CaptureRequest.Builder? = null
//
//    @Mock
//    var surfaceFactoryMock: SurfaceFactory? = null
//
//    @Mock
//    var surfaceMock: Surface? = null
//    var cameraSession: CameraSession? = null
//    var cameraPreview: CameraPreview? = null
//    @Before
//    @Throws(Exception::class)
//    fun setUp() {
//        Mockito.reset(captureRequestBuilder)
//        Mockito.`when`(surfaceFactoryMock.build(ArgumentMatchers.any(Size::class.java)))
//            .thenReturn(surfaceMock)
//        Mockito.`when`(cameraDeviceMock!!.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW))
//            .thenReturn(captureRequestBuilder)
//        cameraSession = CameraSession()
//        cameraPreview = CameraPreview(
//            cameraSession,
//            null,
//            surfaceFactoryMock,
//            null,
//            false
//        )
//        cameraSession!!.onCaptureSessionListenerList = listOf(cameraPreview)
//    }
//
//    @Test
//    @Throws(CameraAccessException::class)
//    fun createPreviewSession() {
//        cameraPreview!!.setPreviewSize(640, 480)
//        cameraPreview!!.createCameraPreviewSession(cameraDeviceMock)
//        Assert.assertNotNull(cameraPreview!!.previewRequest)
//        // Flash is disabled by default
//        Mockito.verify(captureRequestBuilder, Mockito.times(1))
//            .set(CaptureRequest.FLASH_MODE, CaptureRequest.FLASH_MODE_OFF)
//        // AutoFocus is activated by default
//        Mockito.verify(captureRequestBuilder, Mockito.times(1))
//            .set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
//        // Android preview stays on portrait mode
//        Mockito.verify(captureRequestBuilder, Mockito.times(1))
//            .set(CaptureRequest.JPEG_ORIENTATION, 270)
//        // check surfaces have been set
//        Assert.assertEquals(cameraSession!!.surfaces.size.toLong(), 1)
//        Assert.assertEquals(cameraSession!!.cameraDevice, cameraDeviceMock)
//        Assert.assertNotNull(cameraSession!!.captureSession)
//    }
//
//    @Test
//    fun setFocusWithNoPreview() {
//        cameraPreview!!.setAutoFocus(true)
//        Mockito.verify(captureRequestBuilder, Mockito.never()).set(
//            ArgumentMatchers.eq(
//                CaptureRequest.CONTROL_AF_MODE
//            ), Mockito.anyInt()
//        )
//    }
//
//    @Test
//    @Throws(CameraAccessException::class)
//    fun setFocusWithPreview() {
//        cameraPreview!!.setPreviewSize(640, 480)
//        cameraPreview!!.createCameraPreviewSession(cameraDeviceMock)
//        Mockito.reset(captureRequestBuilder)
//        cameraPreview!!.setAutoFocus(true)
//        Mockito.verify(captureRequestBuilder, Mockito.atLeastOnce())
//            .set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
//        Mockito.reset(captureRequestBuilder)
//        cameraPreview!!.setAutoFocus(false)
//        Mockito.verify(captureRequestBuilder, Mockito.atLeastOnce()).set(
//            ArgumentMatchers.eq(
//                CaptureRequest.CONTROL_AF_MODE
//            ), ArgumentMatchers.eq(CameraMetadata.CONTROL_AF_MODE_OFF)
//        )
//    }
//
//    @Test
//    @Throws(CameraAccessException::class)
//    fun setFlashMode() {
//        cameraPreview!!.setPreviewSize(640, 480)
//        cameraPreview!!.createCameraPreviewSession(cameraDeviceMock)
//        Mockito.reset(captureRequestBuilder)
//        cameraPreview!!.setFlashMode(FlashMode.NONE)
//        Mockito.verify(captureRequestBuilder, Mockito.atLeastOnce())
//            .set(CaptureRequest.FLASH_MODE, CaptureRequest.FLASH_MODE_OFF)
//        Mockito.reset(captureRequestBuilder)
//        cameraPreview!!.setFlashMode(FlashMode.AUTO)
//        Mockito.verify(captureRequestBuilder, Mockito.atLeastOnce()).set(
//            ArgumentMatchers.eq(
//                CaptureRequest.FLASH_MODE
//            ), ArgumentMatchers.eq(CameraMetadata.FLASH_MODE_SINGLE)
//        )
//        cameraPreview!!.setFlashMode(FlashMode.ALWAYS)
//        Mockito.verify(captureRequestBuilder, Mockito.atLeastOnce()).set(
//            ArgumentMatchers.eq(
//                CaptureRequest.FLASH_MODE
//            ), ArgumentMatchers.eq(CameraMetadata.FLASH_MODE_TORCH)
//        )
//    }
//}